### DISCUSSION TITLE
API rules for relations

### DISCUSSION THREAD

**Discussion by Siutan** [pocketbase/pocketbase#5604] [Q&A] [answered]
Just wanted to get some insight on the following:

i have a `groups` collection, a `group_members` collection and a `group_messages` collection.
### `groups`:
```
id | name | owner
```

### `group_members`
```
id | group | member | role | status 
```

### `group_messages`:
```
id | group | content | parent_message
```

the issue im having with my rules are that i cannot find a way to:
1. **(list and view)**: Don't show messages if the member has the status `invited`. i tried a few and i have this right now:
```sql
(@request.auth.id ?= group.group_members_via_group.member) && (@collection.group_members.status != 'invited')
```

2. **(create)**: limit creating records in `group_members` to users who are already members and have the role `admin`. ive tried this as well but got nowhere, i had something like this:
```sql
@request.auth.id ?= member
 && 
(@collection.group_members.group.id ?= @request.data.group
  && 
@collection.group_members.member ?= @request.auth.id
  && 
@collection.group_members.role = "admin")
```

im finding the api rules a bit confusing in general, im trying to think of it through a sql lens i guess for example in `point 2`, i would do something like :
```sql
@request.auth.id IN (
  SELECT member
  FROM group_members
  WHERE group = @request.data.group AND role = 'admin'
)
```

but im not sure how to translate that to the pocketbase queries.

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not entirely sure that I understand the SQL pseudo-query for 2, but based on the provided requirements the below API rules should suffice:

1. `group_messages` list/view:
    ```js
    @request.auth.id != "" &&
    group.group_members_via_group.member ?= @request.auth.id &&
    group.group_members_via_group.status ?!= "invited"
    ```

2. `group_members` create:
    ```js
    @request.auth.id != "" &&
    // prevent assigning themself as a member because the Create API rule runs AFTER dry submit of the record
    @request.data.member != @request.auth.id &&
    @collection.group_members.member ?= @request.auth.id &&
    @collection.group_members.group ?= @request.data.group &&
    @collection.group_members.role ?= "admin"
    ```
    OR alternatively using `@request.auth.*` back-relation:
    ```js
    @request.auth.id != "" &&
    // prevent assigning themself as admin member because the Create API rule runs AFTER dry submit of the record
    @request.data.member != @request.auth.id &&
    @request.auth.group_members_via_member.group ?= @request.data.group &&
    @request.auth.group_members_via_member.role ?= "admin"
    ```

The second rule requires the extra `@request.data.member != @request.auth.id` check because currently the Create API rule runs AFTER a dry submit of the record, aka. when performing the self-join the record will aready exist in the `group_members` table.
There are plans to skip the dry create submit but that will be after v0.23 as too many changes ended up in the refactoring.

If the above doesn't work, please export your collections configuration as json (_Admin UI > Settings > Export collections_).

**Reply by ouvreboite** [reply]
@ganigeorgiev looking at this rule

```
@request.auth.id != "" &&
group.group_members_via_group.member ?= @request.auth.id &&
group.group_members_via_group.status ?!= "invited"
```

what would happen if another group’s member has the status `invited`?

I’m reading this `group.group_members_via_group.status ?!= "invited"` as « none of the group’s members have the status invited ».

Or are all the conditions on `group.group_members_via_group` applied together ?

So the overall rule would read like that:
- `group.group_members_via_group.member ?= @request.auth.id` = there must be at least one group’s member with the user id
- `group.group_members_via_group.status ?!= "invited"`  = and none of those ⬆️ group members have the invited status

If that’s the case (conditions on the same relation compound), I’m not sure it’s very explicit in the doc. It’s somehow implied by the `:alias` feature, but I don’t’ think there is anything else.

**Reply by ganigeorgiev** [reply]
Your last suggestion is the correct one.

API rules are essentially evaluated as plain SQL WHERE statements and they execute the same way. An oversimplified pseudo-SQL of the API rule looks something like:

```sql
SELECT DISTINCT *
FROM group_messages
LEFT JOIN groups ON ...
LEFT JOIN group_members ON ...
WHERE {:authId} != "" AND group_members.member = {:authId} AND group_members.status != "invited"
```

Another way to think about the API rules could be as if they were a "funnel" where with each condition you narrow down further the result.

**Reply by ouvreboite** [reply]
Thanks!

That makes it way easier to handle complex use cases :)

**Comment by ElierHG** [comment]
Hello @ganigeorgiev, 

First of all, i want to say thank you for all your hard work and dedication to this amazing project. 

I am reaching out because, after reading the docs and some of these discussion threads, I still don't quite get how relations work in rules. I imagine you must be very tired of answering the same question over and over again, so I am sorry to keep "knocking at your door". 

I would appreciate if you can explain this other scenario:

`users` (basic auth collection)
| id | name | email | etc... |
|--|--|--|--|
| 1 | Test User | Test@gmail.com | |

`groups`:

| id | name | users (relation) |
|---|--------|-------|
| 123 | Sampling Events Admin | Test User

`sampling_events`:

| id | name | created_by | etc... |
|--|--|--|--|
| 1 | Event 1 | ... | ... |

How do I restrict access to the `sampling_events` table based on the group the user belongs to?
I tried:
```javascript
@request.auth.id != "" && 
@collection.groups.name = "Sampling Events Admin" &&
@collection.groups.users.id ?= @request.auth.id
```
and:
```javascript
@request.auth.id != "" && 
@collection.groups.name = "Sampling Events Admin" &&
@collection.groups.users ?= @request.auth.id
```
and many other combinations and could not make it work.

I should clarify that i am using v0.22.14, is not feasible to update my project's version yet.

**Reply by ganigeorgiev** [reply]
> How do I restrict access to the sampling_events table based on the group the user belongs to?

I'm lacking context about your use case to understand the collections structure and why `groups` relates to `sampling_events` through a name field and not via relation (or why the user relation in the `groups` collection is named "users"; is it because it is a multiple relation field?).

In any case if you want to check the current user's group name, then you can try:

```js
@request.auth.id != "" &&
@request.auth.groups_via_users.name ?= "Sampling Events Admin"

// OR manually join with @collection
@request.auth.id != "" &&
@collection.groups.users.id ?= @request.auth.id &&
@collection.groups.name ?= "Sampling Events Admin"
```
_(keep in mind that the name of the collections and fields matter and the above rules will work only if your actual collections and fields are named as you've listed)_

> I should clarify that i am using v0.22.14, is not feasible to update my project's version yet.

Then you should at least try to upgrade to the latest v0.22.x (at the time of writing [v0.22.31](https://github.com/pocketbase/pocketbase/releases/tag/v0.22.31))  since there were several bug fixes, including security related. For more details you can check the [pre v0.23 CHANGELOG](https://github.com/pocketbase/pocketbase/blob/master/CHANGELOG_16_22.md).

**Reply by ElierHG** [reply]
Thank you for replying! Both ways worked perfectly.

I was using the wrong operator (`=` instead of `?=`) when checking the group name. 
```javascript
@collection.groups.name ?= "Sampling Events Admin"
```

Sorry for the lack of context on my previous message. 

I am working in the context of this other open source project called [PocketBlocks ](https://github.com/pedrozadotdev), which combines PocketBase and OpenBlocks (an admin dashboard builder). In that project they use the `groups` collection to share access of front end apps (or dashboards). So, I'm trying to use the same `groups` collection to control access to the API, instead of maintaining a separate one. 

 > why groups relates to `sampling_events` through a name field and not via relation (or why the user relation in the groups collection is named "users"; is it because it is a multiple relation field?)

In the current setup, the `groups` collection has a `users` relation field (one-to-many) instead of the `users` collection having a `groups` field. 

The intent is, if the user belongs to the group called `Sampling Events Admin` they will be granted access to the front end app and the `sampling_events` collection API. 

To be honest, I am not sure if this approach follows best practices or not. 
When it comes to restricting API access, is it better for the `users` collection to have a `groups` relation field or it doesn't matter?
